### PR TITLE
virtiofs: fix virtiofs crash when cache=none

### PR DESCRIPTION
**Pull request by ganeshmaharaj**
When virtio_fs_cache is set to none, the mount options for the folder
inside the guest should not contain the dax option else it leads to
invalid address errors and a crash of the daemon on the host.

Fixes: #1907
Signed-off-by: Ganesh Maharaj Mahalingam <ganesh.mahalingam@intel.com>

### CODE REVIEW

File: virtcontainers/kata_agent.go [hunk 1]
```diff
@@ -66,24 +66,26 @@ var (
...
+	vSockPort                   = 1024
+	kata9pDevType               = "9p"
+	kataMmioBlkDevType          = "mmioblk"
+	kataBlkDevType              = "blk"
+	kataSCSIDevType             = "scsi"
+	kataNvdimmDevType           = "nvdimm"
+	kataVirtioFSDevType         = "virtio-fs"
+	sharedDir9pOptions          = []string{"trans=virtio,version=9p2000.L,cache=mmap", "nodev"}
+	sharedDirVirtioFSOptions    = []string{"default_permissions,allow_other,rootmode=040000,user_id=0,group_id=0,tag=" + mountGuest9pTag, "nodev"}
+	sharedDirVirtioFSDaxOptions = "dax"
```
Comment (devimc) [inline virtcontainers/kata_agent.go]: really hard to identify what was the change here

File: virtcontainers/kata_agent.go [hunk 2]
```diff
@@ -754,6 +756,9 @@ func (k *kataAgent) startSandbox(sandbox *Sandbox) error {
 		// (resolv.conf, etc...) and potentially all container
 		// rootfs will reside.
 		if sandbox.config.HypervisorConfig.SharedFS == config.VirtioFS {
+			if sandbox.config.HypervisorConfig.VirtioFSCache != typeVirtioFSNoCache {
+				sharedDirVirtioFSOptions = append(sharedDirVirtioFSOptions, sharedDirVirtioFSDaxOptions)
```
Comment (devimc) [inline virtcontainers/kata_agent.go]: could you please add a comment here explaining why?

File: virtcontainers/kata_agent.go (hunk 1 above)
Comment (ganeshmaharaj) [reply virtcontainers/kata_agent.go]: I know. I wish golang wouldn't keep moving every line to match every other line. The change is an one line addition.
`sharedDirVirtioFSDaxOptions = "dax"`

Comment (devimc) [reply virtcontainers/kata_agent.go]: and remove `dax` from `sharedDirVirtioFSOptions`

**Comment by ganeshmaharaj** [comment]
@jcvenegas @devimc the initrd fails is the one a revert is going to fix, right?

File: virtcontainers/kata_agent.go [hunk 3]
```diff
@@ -754,6 +756,14 @@ func (k *kataAgent) startSandbox(sandbox *Sandbox) error {
 		// (resolv.conf, etc...) and potentially all container
 		// rootfs will reside.
 		if sandbox.config.HypervisorConfig.SharedFS == config.VirtioFS {
+			// If virtio-fs uses either of the two cache options 'auto, always',
+			// the guest directory can be mounted with option 'dax' allowing it to
+			// directly map contents from the host. When set to 'none', the mount
+			// options should not contain 'dax' lest the virtio-fs daemon crashing
+			// with an invalid address reference.
```
Comment (amshinde) [inline virtcontainers/kata_agent.go]: @ganeshmaharaj Do we know why we get the invalid address reference?

Comment (ganeshmaharaj) [reply virtcontainers/kata_agent.go]: @amshinde cause the guest thinks the memory region is shared between the host daemon and the guest and expects the host to know memory mapping to that region but the host process was started with `cache=none` and does not understand the translation.

**Comment by ganeshmaharaj** [comment]
metrics CI seems to be failing with below error.
```
08:06:18 Error: Error downloading packages:
08:06:18   Curl error (16): Error in the HTTP2 framing layer for https://mirrors.fedoraproject.org/metalink?repo=updates-released-f30&arch=x86_64 []
08:06:18 The command '/bin/sh -c dnf install -y qemu-img parted gdisk e2fsprogs gcc xfsprogs findutils' returned a non-zero code: 1
```
@chavafg any chance you have seen this in the recent past?

**Comment by chavafg** [comment]
not really, maybe a network issue in the infra? I sent a restart, hopefully it doesn't fail this time

**Pull request merged**